[Docs] sylius performance boost (#12873)
Sylius/Sylius: docs/performance/query-optimization.rst

Query optimization
==================

Huge performance boost is achieved by optimizing queries.

For example let's take a look on customers listing.

> By default we dont really need to fetch every single field from customer to create a list of 10 customers,

Reviewer: ```suggestion
By default, we don't really need to fetch every single field from customer to create a list of 10 customers,
```

> all we need is to load id and field we are sorting by (for pagination purposes).

Reviewer: ```suggestion
all we need is to load `id` and field we are sorting by (for pagination purposes).
```

Once these 10 customers are fetched, we can create 10 queries to fetch data for them.
As it can slow down small databases (a lot of unnecessary queries), this allows tables with millions of records to be loaded in less than a second.

> It seems to be a bit complicated, but DoctrineORMAdapter allows us to achieve easily.

Reviewer: ```suggestion
It seems to be a bit complicated, but `DoctrineORMAdapter` allows us to achieve easily.
```

> It seems to be a bit complicated, but `DoctrineORMAdapter` allows us to achieve easily.

Reviewer 2: ```suggestion
It seems to be a bit complicated, but `DoctrineORMAdapter` allows us to achieve it easily.
```

> Second argument of DoctrineORMAdapter is fetchJoinCollection, which is false by default.

Reviewer: ```suggestion
Second argument of `DoctrineORMAdapter` is `fetchJoinCollection`, which is false by default.
```

> Second argument of `DoctrineORMAdapter` is `fetchJoinCollection`, which is false by default.

Reviewer 2: ```suggestion
The second argument of `DoctrineORMAdapter` is `fetchJoinCollection`, which is set to `false` by default.
```

> Changing it to true, forces database to fetch additional data once we get sorted results.

Reviewer 2: ```suggestion
Changing it to `true` forces the database to fetch additional data once we get sorted results.
```

With 3 000 000 customers this method allows you to load page up to 70% faster.

.. warning::

>     This solution may slow down loading page with small tables as it will create additional database queries.

Reviewer 3: Can we also point out how to achieve it? At least mention that it will be possible from the version X of GridBundle

Query optimization
==================

> Huge performance boost is achieved by optimizing queries.

Reviewer 3: ```suggestion
A huge performance boost is achieved by optimizing queries.
```

> For example let's take a look on customers listing.

Reviewer 3: ```suggestion
For example, let's take a look at the customers' list.
```

> By default, we don't really need to fetch every single field from customer to create a list of 10 customers,

Reviewer 3: ```suggestion
By default, we don't really need to fetch every single field from a customer to create a list of 10 customers.
```

> all we need is to load `id` and field we are sorting by (for pagination purposes).

Reviewer 3: ```suggestion
All we need, is to load the `id` and the field we are sorting by (for pagination purposes).
```

Once these 10 customers are fetched, we can create 10 queries to fetch data for them.
As it can slow down small databases (a lot of unnecessary queries), this allows tables with millions of records to be loaded in less than a second.

> It seems to be a bit complicated, but `DoctrineORMAdapter` allows us to achieve it easily.

Reviewer 3: ```suggestion
It seems to be a bit complicated, but `Pagerfanta\Adapter\DoctrineORMAdapter` allows us to achieve it easily.
```

> The second argument of `DoctrineORMAdapter` is `fetchJoinCollection`, which is set to `false` by default.

Reviewer 3: ```suggestion
The second argument of `Pagerfanta\Adapter\DoctrineORMAdapter` is `fetchJoinCollection`, which is set to `false` by default.
```

Changing it to `true`, forces the database to fetch additional data once we get sorted results.

With 3 000 000 customers this method allows you to load page up to 70% faster.

.. warning::

    This solution may slow down loading page with small tables as it will create additional database queries.

> This solution will be available on `SyliusGridBundle` since version 1.10.

Reviewer 3: ```suggestion
This solution will be turned on by default on `SyliusGridBundle` since version 1.10. For the earlier versions, the `\Pagerfanta\Adapter\DoctrineORMAdapter` has to be adjusted manually, by overriding the `\Sylius\Bundle\GridBundle\Doctrine\ORM\DataSource` class.
```

A huge performance boost is achieved by optimizing queries.
For example, let's take a look at the customers' list.

By default, we don't really need to fetch every single field from a customer entity to create a list of 10 customers.
All we need, is to load the `id` and the field we are sorting by (for pagination purposes).

Once these 10 customers are fetched, we can create 10 queries to fetch data for them.
As it can slow down small databases (a lot of unnecessary queries), this allows tables with millions of records to be loaded in less than a second.

It seems to be a bit complicated, but `Pagerfanta\Adapter\DoctrineORMAdapter` allow us to achieve it easily.

The second argument of `Pagerfanta\Adapter\DoctrineORMAdapter` is `fetchJoinCollection`, which is set to `false` by default.
Changing it to `true`, forces the database to fetch additional data once we get sorted results.

With 3 000 000 customers this method allows you to load page up to 70% faster.

.. warning::

    This solution may slow down loading page with small tables as it will create additional database queries.

This solution will be turned on by default on `SyliusGridBundle` since version 1.10.

> For the earlier versions, the `\Pagerfanta\Adapter\DoctrineORMAdapter` has to be adjusted manually,
> by overriding the `\Sylius\Bundle\GridBundle\Doctrine\ORM\DataSource` class.

Reviewer 4: ```suggestion
For the earlier versions, the `Pagerfanta\Adapter\DoctrineORMAdapter` has to be adjusted manually,
by overriding the `Sylius\Bundle\GridBundle\Doctrine\ORM\DataSource` class.
```
